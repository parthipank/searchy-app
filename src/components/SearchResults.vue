<template>
  <div class="wrapper">
    <div
      class="item"
      v-for="(item, name) in searchResultItem"
      v-bind:key="item + name"
    >
      <p class="result-data">
        <b>{{ name }}</b> : {{ item }}
      </p>
    </div>
  </div>
</template>

<script>
export default {
  name: "SearchResults",
  props: ["searchResultItem"],
};
</script>

<style scoped>
.item {
  text-align: left;
  padding-left: 10px;
  padding-right: 10px;
}
.wrapper {
  text-align: left;
  background: lightgrey;
  border-color: lightslategrey;
  margin-left: 50px;
  margin-right: 50px;
}
</style>
