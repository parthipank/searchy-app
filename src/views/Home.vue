<template>
  <div id="app">
    <Header />
    <Search v-on:search="onSearch" />
    <div
      v-for="(searchResultItem, index) in searchResults"
      v-bind:key="searchResultItem + index"
    >
      <SearchResults v-bind:searchResultItem="searchResultItem" />
    </div>
  </div>
</template>

<script>
import Search from "../components/Search";
import SearchResults from "../components/SearchResults";
import organizations from "../store/organizations.json";
import users from "../store/users.json";
import tickets from "../store/tickets.json";

export default {
  name: "Home",
  components: {
    Search,
    SearchResults,
  },
  data() {
    return {
      organizations,
      tickets,
      users,
      searchResults: [],
    };
  },
  methods: {
    onSearch(searchData) {
      this.searchResults = [];
      if (searchData.entity == "organizations") {
        organizations.map((org) => {
          if (
            searchData.term === "domain_names" ||
            searchData.term === "tags"
          ) {
            org[searchData.term].map((item) => {
              if (item.toString() === searchData.search) {
                this.tickUsers(tickets, users, org);
              }
            });
          }
          if (
            org[searchData.term].toString() &&
            org[searchData.term].toString() === searchData.search
          ) {
            this.tickUsers(tickets, users, org);
          }
        });
      }
      if (searchData.entity == "tickets") {
        tickets.map((ticket) => {
          if (searchData.term === "tags") {
            ticket[searchData.term].map((item) => {
              if (item.toString() === searchData.search) {
                this.orgUsers(organizations, users, ticket);
              }
            });
          }
          if (
            ticket[searchData.term].toString() &&
            ticket[searchData.term].toString() === searchData.search
          ) {
            this.orgUsers(organizations, users, ticket);
          }
        });
      }
      if (searchData.entity == "users") {
        users.map((user) => {
          if (searchData.term === "tags") {
            user[searchData.term].map((item) => {
              if (item.toString() === searchData.search) {
                this.tickOrgs(tickets, organizations, user);
              }
            });
          }
          if (
            user[searchData.term] &&
            user[searchData.term].toString() === searchData.search
          ) {
            this.tickOrgs(tickets, organizations, user);
          }
        });
      }
      if (this.searchResults.length === 0) {
        alert("No Search Results, Make your search key precise");
      }
    },
    tickUsers(tickets, users, org) {
      let tempObj = {};

      tempObj["name"] = org.name;
      tempObj["url"] = org.url;
      tempObj["external_id"] = org.external_id;
      tempObj["domain_names"] = org.details.toString();
      tempObj["created_at"] = org.created_at;
      tempObj["details"] = org.details;
      tempObj["shared_tickets"] = org.shared_tickets.toString();
      tempObj["tags"] = org.tags.toString();

      let t = 0;
      tickets.map((ticket) => {
        if (ticket.organization_id === org._id) {
          tempObj["ticket_" + t] = ticket.subject;
          t++;
        }
      });

      let u = 0;
      users.map((user) => {
        if (user.organization_id === org._id) {
          tempObj["user_" + u] = user.name;
          u++;
        }
      });
      this.searchResults.push(tempObj);
    },
    orgUsers(organizations, users, ticket) {
      let tempObj = {};
      tempObj["subject"] = ticket.subject;
      tempObj["url"] = ticket.url;
      tempObj["external_id"] = ticket.external_id;
      tempObj["type"] = ticket.type;
      tempObj["description"] = ticket.description;
      tempObj["priority"] = ticket.priority;
      tempObj["status"] = ticket.status;
      tempObj["tags"] = ticket.tags.toString();
      tempObj["has_incidents"] = ticket.has_incidents.toString();
      tempObj["due_at"] = ticket.due_at;
      tempObj["via"] = ticket.via;

      organizations.map((org) => {
        if (ticket.organization_id === org._id) {
          tempObj["Organization Name"] = org.name;
        }
      });

      users.map((user) => {
        if (user._id === ticket.assignee_id) {
          tempObj["Assignee Name"] = user.name;
        }
        if (user._id === ticket.submitter_id) {
          tempObj["Submitter Name"] = user.name;
        }
      });
      this.searchResults.push(tempObj);
    },
    tickOrgs(tickets, organizations, user) {
      let tempObj = {};

      tempObj["name"] = user.name;
      tempObj["alias"] = user.alias;
      tempObj["external_id"] = user.external_id;
      tempObj["role"] = user.role;
      tempObj["signature"] = user.signature;
      tempObj["suspended"] = user.suspended.toString();
      tempObj["verified"] = user.verified.toString();
      tempObj["active"] = user.active.toString();
      tempObj["tags"] = user.tags.toString();
      tempObj["url"] = user.url;

      tickets.map((ticket) => {
        if (user._id === ticket.assignee_id) {
          tempObj["Assignee Ticket"] = ticket.subject;
        }
        if (user._id === ticket.submitter_id) {
          tempObj["Submitter Ticket"] = ticket.subject;
        }
      });

      organizations.map((org) => {
        if (user.organization_id === org._id) {
          tempObj["Organization Name"] = org.name;
        }
      });
      this.searchResults.push(tempObj);
    },
  },
};
</script>

<style>
#app {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-align: center;
  color: #2c3e50;
}

body {
  font-family: Arial, Helvetica, sans-serif;
  line-height: 1.4;
}
</style>
