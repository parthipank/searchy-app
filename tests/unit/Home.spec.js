import { mount } from "@vue/test-utils";
import Home from "../../src/views/Home";
import SearchResults from "../../src/components/SearchResults";

it("Expect desired search result", () => {
  const wrapper = mount(Home);
  //Mock Data
  wrapper.setData({
    organizations: [
      {
        _id: 102,
        url: "http://initech.tokoin.io.com/api/v2/organizations/102.json",
        external_id: "7cd6b8d4-2999-4ff2-8cfd-44d05b449226",
        name: "Nutralab",
        domain_names: [
          "trollery.com",
          "datagen.com",
          "bluegrain.com",
          "dadabase.com",
        ],
        created_at: "2016-04-07T08:21:44 -10:00",
        details: "Non profit",
        shared_tickets: false,
        tags: ["Cherry", "Collier", "Fuentes", "Trevino"],
      },
    ],
  });
  wrapper.find("#entity").setValue("organizations");
  wrapper.find("#term").setValue("name");
  wrapper.find("#searchvalue").setValue("Nutralab");
  wrapper.find("button").trigger("click");
  // expect(wrapper.find({ name: "SearchResults" }).exists()).toBe(true);

  expect(wrapper.element).toMatchSnapshot();
});
