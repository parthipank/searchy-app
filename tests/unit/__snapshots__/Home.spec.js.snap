// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expect desired search result 1`] = `
<div
  id="app"
>
  <header />
   
  <div>
    <form>
      <label
        for="entity"
      >
        Search Entity 
      </label>
       
      <select
        class="termbox"
        id="entity"
      >
        <option
          value="users"
        >
          Users
        </option>
         
        <option
          value="organizations"
        >
          Organizations
        </option>
         
        <option
          value="tickets"
        >
          Tickets
        </option>
      </select>
       
      <label
        for="term"
      >
         Search Term
      </label>
       
      <select
        class="termbox"
        id="term"
      >
        <option
          value="_id"
        >
          _id
        </option>
        <option
          value="url"
        >
          url
        </option>
        <option
          value="external_id"
        >
          external_id
        </option>
        <option
          value="name"
        >
          name
        </option>
        <option
          value="alias"
        >
          alias
        </option>
        <option
          value="created_at"
        >
          created_at
        </option>
        <option
          value="active"
        >
          active
        </option>
        <option
          value="verified"
        >
          verified
        </option>
        <option
          value="shared"
        >
          shared
        </option>
        <option
          value="locale"
        >
          locale
        </option>
        <option
          value="timezone"
        >
          timezone
        </option>
        <option
          value="last_login_at"
        >
          last_login_at
        </option>
        <option
          value="email"
        >
          email
        </option>
        <option
          value="phone"
        >
          phone
        </option>
        <option
          value="signature"
        >
          signature
        </option>
        <option
          value="organization_id"
        >
          organization_id
        </option>
        <option
          value="tags"
        >
          tags
        </option>
        <option
          value="suspended"
        >
          suspended
        </option>
        <option
          value="role"
        >
          role
        </option>
      </select>
       
      <input
        class="termbox"
        id="searchvalue"
        name="searchvalue"
        type="text"
      />
       
      <button
        class="btn"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
   
</div>
`;
